File: api/tools/src/main/java/org/pmiops/workbench/tools/BackfillBillingProjectUsers.java
package org.pmiops.workbench.tools;

import com.google.common.collect.ImmutableList;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import java.lang.reflect.Type;
import java.util.List;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.DefaultParser;
import org.apache.commons.cli.Option;
import org.apache.commons.cli.Options;
import org.pmiops.workbench.firecloud.ApiException;
import org.pmiops.workbench.firecloud.api.BillingApi;
import org.pmiops.workbench.firecloud.api.WorkspacesApi;
import org.pmiops.workbench.firecloud.model.FirecloudWorkspace;
import org.pmiops.workbench.firecloud.model.FirecloudWorkspaceACL;
import org.pmiops.workbench.firecloud.model.FirecloudWorkspaceAccessEntry;
import org.pmiops.workbench.firecloud.model.FirecloudWorkspaceResponse;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.builder.SpringApplicationBuilder;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

/**
 * Backfill script for granting the billing project user role to all AoU owners (corresponding to
 * Firecloud OWNER access). For http://broad.io/1ppw, collaborators need to be either writers or
 * billing project users in order to launch clusters within shared billing projects, see RW-3009 and
 * RW-3188 for details.
 */
@Configuration
public class BackfillBillingProjectUsers {
  public static final List<String> FIRECLOUD_LIST_WORKSPACES_REQUIRED_FIELDS =
      ImmutableList.of(
          "accessLevel", "workspace.namespace", "workspace.name", "workspace.createdBy");

  private static Option fcBaseUrlOpt =
      Option.builder()
          .longOpt("fc-base-url")
          .desc("Firecloud API base URL")
          .required()
          .hasArg()
          .build();
  private static Option billingProjectPrefixOpt =
      Option.builder()
          .longOpt("billing-project-prefix")
          .desc("Billing project prefix to filter by, other workspaces are ignored")
          .required()
          .hasArg()
          .build();
  private static Option dryRunOpt =
      Option.builder()
          .longOpt("dry-run")
          .desc("If specified, the tool runs in dry run mode; no modifications are made")
          .build();
  private static Options options =
      new Options().addOption(fcBaseUrlOpt).addOption(billingProjectPrefixOpt).addOption(dryRunOpt);

  private static final Logger log = Logger.getLogger(BackfillBillingProjectUsers.class.getName());

  private static void dryLog(boolean dryRun, String msg) {
    String prefix = "";
    if (dryRun) {
      prefix = "[DRY RUN] Would have... ";
    }
    log.info(prefix + msg);
  }

  /**
   * Swagger Java codegen does not handle the WorkspaceACL model correctly; it returns a GSON map
   * instead. Run this through a typed Gson conversion process to coerce it into the desired type.
   */
  public static Map<String, FirecloudWorkspaceAccessEntry> extractAclResponse(
      FirecloudWorkspaceACL aclResp) {
    Type accessEntryType = new TypeToken<Map<String, FirecloudWorkspaceAccessEntry>>() {}.getType();
    Gson gson = new Gson();
    return gson.fromJson(gson.toJson(aclResp.getAcl(), accessEntryType), accessEntryType);
  }

  private static void backfill(
      WorkspacesApi workspacesApi,
      BillingApi billingApi,
      String billingProjectPrefix,
      boolean dryRun)
      throws ApiException {
    int userUpgrades = 0;
    for (FirecloudWorkspaceResponse resp :
        workspacesApi.listWorkspaces(FIRECLOUD_LIST_WORKSPACES_REQUIRED_FIELDS)) {
      FirecloudWorkspace w = resp.getWorkspace();
      if (!w.getNamespace().startsWith(billingProjectPrefix)) {
        continue;
      }

      String id = w.getNamespace() + "/" + w.getName();
      if (!"PROJECT_OWNER".equals(resp.getAccessLevel())) {
        log.warning(
            String.format(
                "service account has '%s' access to workspace '%s'; skipping",
                resp.getAccessLevel(), id));
        continue;
      }

      Map<String, FirecloudWorkspaceAccessEntry> acl =
          extractAclResponse(workspacesApi.getWorkspaceAcl(w.getNamespace(), w.getName()));
      for (String user : acl.keySet()) {
        if (user.equals(w.getCreatedBy())) {
          // Skip the common case, creators should already be billing project users. Any edge cases
          // where this is not true will be fixed by the 1PPW migration (RW-2705).
          continue;
        }
        FirecloudWorkspaceAccessEntry entry = acl.get(user);
        if (!"OWNER".equals(entry.getAccessLevel())) {
          // Only owners should be granted billing project user.
          continue;
        }
        dryLog(
            dryRun,
            String.format(
                "granting billing project user on '%s' to '%s' (%s)",
                w.getNamespace(), user, entry.getAccessLevel()));
        if (!dryRun) {
          try {
            billingApi.addUserToBillingProject(w.getNamespace(), "user", user);
          } catch (ApiException e) {
            log.log(Level.WARNING, "failed to add user to project", e);
          }
        }
        userUpgrades++;
      }
    }

    dryLog(dryRun, String.format("added %d users as billing project users", userUpgrades));
  }

  @Bean
  public CommandLineRunner run() {
    return (args) -> {
      CommandLine opts = new DefaultParser().parse(options, args);

      ServiceAccountAPIClientFactory apiFactory =
          new ServiceAccountAPIClientFactory(opts.getOptionValue(fcBaseUrlOpt.getLongOpt()));

      backfill(
          apiFactory.workspacesApi(),
          apiFactory.billingApi(),
          opts.getOptionValue(billingProjectPrefixOpt.getLongOpt()),
          opts.hasOption(dryRunOpt.getLongOpt()));
    };
  }

  public static void main(String[] args) {
    new SpringApplicationBuilder(BackfillBillingProjectUsers.class).web(false).run(args);
  }
}

File: api/tools/src/main/java/org/pmiops/workbench/tools/BackfillBillingProjectUsers.kt
package org.pmiops.workbench.tools

import com.google.api.client.googleapis.auth.oauth2.GoogleCredential
import com.google.common.collect.ImmutableList
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.io.IOException
import java.lang.reflect.Type
import java.util.Arrays
import java.util.logging.Level
import java.util.logging.Logger
import org.apache.commons.cli.CommandLine
import org.apache.commons.cli.DefaultParser
import org.apache.commons.cli.Option
import org.apache.commons.cli.Options
import org.pmiops.workbench.firecloud.ApiClient
import org.pmiops.workbench.firecloud.ApiException
import org.pmiops.workbench.firecloud.api.BillingApi
import org.pmiops.workbench.firecloud.api.WorkspacesApi
import org.pmiops.workbench.firecloud.model.Workspace
import org.pmiops.workbench.firecloud.model.WorkspaceACL
import org.pmiops.workbench.firecloud.model.WorkspaceAccessEntry
import org.pmiops.workbench.firecloud.model.WorkspaceResponse
import org.springframework.boot.CommandLineRunner
import org.springframework.boot.builder.SpringApplicationBuilder
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

/**
 * Backfill script for granting the billing project user role to all AoU owners (corresponding to
 * Firecloud OWNER access). For http://broad.io/1ppw, collaborators need to be either writers or
 * billing project users in order to launch clusters within shared billing projects, see RW-3009 and
 * RW-3188 for details.
 */
@Configuration
class BackfillBillingProjectUsers {

    @Bean
    fun run(): CommandLineRunner {
        return { args ->
            val opts = DefaultParser().parse(options, args)
            backfill(
                    newWorkspacesApi(opts.getOptionValue(fcBaseUrlOpt.longOpt)),
                    newBillingApi(opts.getOptionValue(fcBaseUrlOpt.longOpt)),
                    opts.getOptionValue(billingProjectPrefixOpt.longOpt),
                    opts.hasOption(dryRunOpt.longOpt))
        }
    }

    companion object {
        val FIRECLOUD_LIST_WORKSPACES_REQUIRED_FIELDS: List<String> = ImmutableList.of(
                "accessLevel", "workspace.namespace", "workspace.name", "workspace.createdBy")

        private val fcBaseUrlOpt = Option.builder()
                .longOpt("fc-base-url")
                .desc("Firecloud API base URL")
                .required()
                .hasArg()
                .build()
        private val billingProjectPrefixOpt = Option.builder()
                .longOpt("billing-project-prefix")
                .desc("Billing project prefix to filter by, other workspaces are ignored")
                .required()
                .hasArg()
                .build()
        private val dryRunOpt = Option.builder()
                .longOpt("dry-run")
                .desc("If specified, the tool runs in dry run mode; no modifications are made")
                .build()
        private val options = Options().addOption(fcBaseUrlOpt).addOption(billingProjectPrefixOpt).addOption(dryRunOpt)

        private val FC_SCOPES = arrayOf("https://www.googleapis.com/auth/userinfo.profile", "https://www.googleapis.com/auth/userinfo.email")

        private val log = Logger.getLogger(BackfillBillingProjectUsers::class.java.name)

        @Throws(IOException::class)
        private fun newApiClient(apiUrl: String): ApiClient {
            val apiClient = ApiClient()
            apiClient.setBasePath(apiUrl)
            val credential = GoogleCredential.getApplicationDefault().createScoped(Arrays.asList(*FC_SCOPES))
            credential.refreshToken()
            apiClient.setAccessToken(credential.accessToken)
            return apiClient
        }

        @Throws(IOException::class)
        private fun newWorkspacesApi(apiUrl: String): WorkspacesApi {
            val api = WorkspacesApi()
            api.setApiClient(newApiClient(apiUrl))
            return api
        }

        @Throws(IOException::class)
        private fun newBillingApi(apiUrl: String): BillingApi {
            val api = BillingApi()
            api.setApiClient(newApiClient(apiUrl))
            return api
        }

        private fun dryLog(dryRun: Boolean, msg: String) {
            var prefix = ""
            if (dryRun) {
                prefix = "[DRY RUN] Would have... "
            }
            log.info(prefix + msg)
        }

        /**
         * Swagger Java codegen does not handle the WorkspaceACL model correctly; it returns a GSON map
         * instead. Run this through a typed Gson conversion process to coerce it into the desired type.
         */
        private fun extractAclResponse(aclResp: WorkspaceACL): Map<String, WorkspaceAccessEntry> {
            val accessEntryType = object : TypeToken<Map<String, WorkspaceAccessEntry>>() {

            }.type
            val gson = Gson()
            return gson.fromJson(gson.toJson(aclResp.getAcl(), accessEntryType), accessEntryType)
        }

        @Throws(ApiException::class)
        private fun backfill(
                workspacesApi: WorkspacesApi,
                billingApi: BillingApi,
                billingProjectPrefix: String,
                dryRun: Boolean) {
            var userUpgrades = 0
            for (resp in workspacesApi.listWorkspaces(FIRECLOUD_LIST_WORKSPACES_REQUIRED_FIELDS)) {
                val w = resp.getWorkspace()
                if (!w.getNamespace().startsWith(billingProjectPrefix)) {
                    continue
                }

                val id = w.getNamespace() + "/" + w.getName()
                if ("PROJECT_OWNER" != resp.getAccessLevel()) {
                    log.warning(
                            String.format(
                                    "service account has '%s' access to workspace '%s'; skipping",
                                    resp.getAccessLevel(), id))
                    continue
                }

                val acl = extractAclResponse(workspacesApi.getWorkspaceAcl(w.getNamespace(), w.getName()))
                for (user in acl.keys) {
                    if (user == w.getCreatedBy()) {
                        // Skip the common case, creators should already be billing project users. Any edge cases
                        // where this is not true will be fixed by the 1PPW migration (RW-2705).
                        continue
                    }
                    val entry = acl[user]
                    if ("OWNER" != entry.getAccessLevel()) {
                        // Only owners should be granted billing project user.
                        continue
                    }
                    dryLog(
                            dryRun,
                            String.format(
                                    "granting billing project user on '%s' to '%s' (%s)",
                                    w.getNamespace(), user, entry.getAccessLevel()))
                    if (!dryRun) {
                        try {
                            billingApi.addUserToBillingProject(w.getNamespace(), "user", user)
                        } catch (e: ApiException) {
                            log.log(Level.WARNING, "failed to add user to project", e)
                        }

                    }
                    userUpgrades++
                }
            }

            dryLog(dryRun, String.format("added %d users as billing project users", userUpgrades))
        }

        @JvmStatic
        fun main(args: Array<String>) {
            SpringApplicationBuilder(BackfillBillingProjectUsers::class.java).web(false).run(*args)
        }
    }
}
